### DISCUSSION TITLE
What does `ISMAGS.analyze_symmetry` do?

### DISCUSSION THREAD

**Discussion by attila-i-szabo** [networkx/networkx#4761] [Q&A]
Hi, I'm trying to use NetworkX to construct the automorphism group of a graph in an "intelligible" form, i.e. try to find a small number of generators that would generate the whole group. I see that `ISMAGS` is able to generate the full automorphism group, and saw that the `analyze_symmetry` method promises to return a "minimal set of permutations" which sounds very much like what I want. However, the documentation says hardly anything about what it actually does and what the parameters should be. Could you please help explain what this method does? Thanks!

**Comment by dschult** [comment]
The [documentation ](https://networkx.org/documentation/stable/reference/algorithms/isomorphism.ismags.html) contains two citations. Probably the link to the original article that introduced ISMAGS is most appropriate.

**Reply by attila-i-szabo** [reply]
Sure, this actually sounds like what I want. Still, I can't figure out what the parameters `node_partitions` and `edge_colors` mean, and the paper doesn't help with this question, nor does the documentation. For instance, what should I pass for these arguments if I only want a representation of the full automorphism group of the graph (without any coloured or directed edges or anything like that) I pass as the first argument?

**Reply by dschult** [reply]
I see the question now...  That documentation is really poorly implemented. 

As far as I can tell the ```analyze_symmetry``` method is called by the ```find_isomorphism``` and ```largest_common_subgraph``` methods. Both use the input call: `analyze_symmetry(self.subgraph, self._sgn_partitions, self._sge_colors)`.  So that answers a small part of your question. But probably isn't all that helpful.  It looks to me like this method isn't prepared to be publicly used and maybe some of the others also.  To really understand the implementation here fully it looks like the source-code is better than the docs *sigh*. 

But here's an overview. To get he automorphisms, you should create an instance of the `ISMAGS` class with the full graph entered as both the graph and the subgraph.  You'll need to provide boolean functions to determine whether two nodes should be considered the same and whether two edges should be considered the same. If you don't have categories of nodes or edges these can both be:   `lambda g1,x,g2,y: True`.  Then the method `find_isomorphisms` is probably what you want. I suggest trying it out on a small graph first so the response is quick and you can understand input and output.  Note that `find_isomorphisms` uses the `analyze_symmetry` method so you don't have to.

As noted in the docs, the VF2 isomorphism routines are likely to be faster, but ISMAGS may be better about giving a reduced set that can generate the other isomorphorisms.
